cli/lock.ml

```diff
@@ -192,9 +192,34 @@ let display_verbose_diagnostics = function
   | None -> false
   | Some l -> l >= Logs.Info
 
+let could_not_determine_version offending_packages =
+  let f (relop, version) =
+    let pp_relop fmt = function
+      | `Eq -> Fmt.pf fmt "="
+      | `Geq -> Fmt.pf fmt ">="
+      | `Gt -> Fmt.pf fmt ">"
+      | `Leq -> Fmt.pf fmt "<="
+      | `Lt -> Fmt.pf fmt "<"
+      | `Neq -> Fmt.pf fmt "!="
+    in
+    Fmt.str "%a %a" pp_relop relop Opam.Pp.version version
+  in
+  let s = OpamFormula.string_of_formula f in
+  let pp_version_formula = Fmt.using s Fmt.string in
+  let pp_offending_package =
+    Fmt.pair ~sep:Fmt.sp Opam.Pp.package_name pp_version_formula
+  in
+  let pp_offending_packages = Fmt.list ~sep:Fmt.comma pp_offending_package in
+  Logs.err (fun l ->
+      l "There is no eligible package that matches %a." pp_offending_packages
```

Comment:
I wonder if this wouldn't be more accurate and clear if it said "There is no eligible version of <pkgname> that matches <constraint>".
We would have to print this for each offending package but I think it would make the whole thing much clearer.

With the current situation if there are several offending packages, it would read like this:
```
There is no eligible package that matches a >= 1.0, b < 0.5
```
and it's not entirely clear whether it's one package for each or one for all from the phrasing which I think can make things confusing to users.

Reply:
This is indeed much better. I think it reveals that I only ever tested it with one package that doesn't match so the wording was all wrong, while the code (and reality) supports more than one thing being wrong. Fixed this now, I it is much better.